test/common/network/happy_eyeballs_connection_impl_test.cc

```diff
@@ -630,13 +633,25 @@ TEST_F(HappyEyeballsConnectionImplTest, WriteBeforeConnectOverLimitWithCallbacks
   impl_->addConnectionCallbacks(callbacks);
 
   Buffer::OwnedImpl data("hello world");
+  size_t length = data.length();
   bool end_stream = false;
 
-  EXPECT_CALL(*created_connections_[0], setBufferLimits(data.length() - 1));
+  EXPECT_CALL(*created_connections_[0], setBufferLimits(length - 1));
   impl_->setBufferLimits(data.length() - 1);
 
   EXPECT_CALL(callbacks, onAboveWriteBufferHighWatermark());
   impl_->write(data, end_stream);
+
+  {
+    testing::InSequence s;
+    // The call to write() will be replayed on the underlying connection.
+    EXPECT_CALL(*created_connections_[0], removeConnectionCallbacks(_));
+    EXPECT_CALL(*failover_timer_, disableTimer());
+    EXPECT_CALL(*created_connections_[0], bufferLimit()).WillRepeatedly(Return(length - 1));
+    EXPECT_CALL(*created_connections_[0], write(_, _));
```

Comment:
can we have this Invoke high watermark callbacks and make sure it isn't passed up to the original user?

then after the Connected call, invoke low watermark and make sure it is passed up

Reply:
I've added a comment here to hopefully explain what's going on. Basically we replay the write when there are no connection callbacks set on the connection ("remove" then "write" then "add"). This means that the above signal is swallowed *in* the underlying connection. Does that sound reasonable?